patches/playroom+0.28.0.patch

```diff
@@ -163,46 +170,76 @@ index 9285b94..c268bdb 100644
    const iframeRef = useRef<HTMLIFrameElement | null>(null);
    const intersection = useIntersection(iframeRef, {
      root: intersectionRootRef.current,
-@@ -32,35 +39,41 @@ export default function Iframe({
+@@ -29,8 +36,17 @@ export default function Iframe({
+     threshold: 0,
+   });
+ 
++  const postMessageToIframe = useCallback((message) => {
++    // Note that we're sending the message to '*', rather than some
++    // specific origin. Sandboxed iframes without the 'allow-same-origin'
++    // header don't have an origin which we can target.
++    // See: https://web.dev/articles/sandboxed-iframes
++    iframeRef.current?.contentWindow?.postMessage(message, '*');
++  }, []);
++
    const intersectionRatio = intersection?.intersectionRatio ?? null;
  
++  // Only render the iframe once it's actually on screen
    useEffect(() => {
--    if (intersectionRatio === null) {
--      return;
-+    if (intersectionRatio === null) return;
-+    const listener = (e: any) => {
-+      if (e.data.type === READY_TO_RECEIVE && e.source === iframeRef.current?.contentWindow) {
-+        setReadyToPost(true);
-+        const contentWindow = iframeRef.current?.contentWindow;
-+        if (contentWindow) {
-+          contentWindow.postMessage({ type: INCOMING_CODE,  themeName, code }, window.origin);
-+        }
-+      }
-     }
-+    window.addEventListener('message', listener);
- 
-     if (intersectionRatio > 0 && src !== renderedSrc) {
-       setRenderedSrc(src);
+     if (intersectionRatio === null) {
+       return;
+@@ -41,26 +57,56 @@ export default function Iframe({
      }
-+    return () => window.removeEventListener('message', listener);
    }, [intersectionRatio, src, renderedSrc]);
  
++  // A call-and-response / ping-pong to wait for the iframe to declare itself
++  // ready. Why do we do polling instead of just having the iframe send a
++  // message to `window.origin`? Because sandboxded iframes without the
++  // 'allow-same-origin' header don't have an origin which it can target. So
++  // instead we send messages to the iframe which can then use `event.origin`.
    useEffect(() => {
 -    if (renderedSrc !== null) {
 -      const location = iframeRef.current?.contentWindow?.location;
--
++    // If the iframe isn't on screen / loaded yet, no need to poll yet.
++    // If the iframe is already ready, don't do anything.
++    if (!loaded || readyToPost) {
++      return;
++    }
+ 
 -      if (location) {
 -        location.replace(renderedSrc);
```

Comment:
The original patch here was to work around a URL length limit when setting the location to include the complete, un-encoded source (when large code examples were written in the Playroom editor). This was achieved by using `window.postMessage` to send the required code to the iframe.

The iframe would send a message to the parent frame (`window.parent.postMessage`) once it's ready to start receiving code change messages. We do this to avoid the frame missing messages about code changes.

However, when the iframe is rendered within a sandoxed environment without `allow-cross-origin`, `window.parent` is undefined, as-is `window.origin`. Instead, we employee a polling strategy where the parent frame polls the iframe with `postMessage()`s. Once the frame is setup and ready to go, it leverages the `postMessage#event` param to figure out the parent & origins for sending the reply.